File: Sources/AstrolabeProtocol/Negotiation/RuntimeHandshake.swift
//
//  RuntimeHandshake.swift
//  astrolabe-protocol
//
//  Created by 轩辕十四 on 2026/7/16.
//

public struct RuntimeClientDescriptor: Codable, Equatable, Sendable {
    /// Client product name.
    public let name: String

    /// Client release version.
    public let version: String

    public init(name: String, version: String) {
        self.name = name
        self.version = version
    }
}

public struct RuntimeHandshakeParameters: Codable, Equatable, Sendable, RuntimeMethodBound {
    /// Host initiating the protocol session.
    public let client: RuntimeClientDescriptor

    /// Protocol versions the Host can decode.
    public let supportedProtocolRange: RuntimeProtocolRange

    public init(client: RuntimeClientDescriptor, supportedProtocolRange: RuntimeProtocolRange) {
        self.client = client
        self.supportedProtocolRange = supportedProtocolRange
    }

    public static let runtimeMethod = RuntimeMethod.handshake
}

public struct RuntimeDescriptor: Codable, Equatable, Sendable {
    /// Namespaced runtime implementation identifier.
    public let identifier: RuntimeNamespacedIdentifier

    /// Runtime SDK release version.
    public let version: String

    /// Opaque identifier for this runtime process instance.
    public let instanceID: RuntimeOpaqueIdentifier

    public init(
        identifier: RuntimeNamespacedIdentifier,
        version: String,
        instanceID: RuntimeOpaqueIdentifier
    ) {
        self.identifier = identifier
        self.version = version
        self.instanceID = instanceID
    }
}

public struct RuntimeHandshakePayload: Codable, Equatable, Sendable, RuntimeMethodBound {
    /// Runtime implementation participating in the session.
    public let runtime: RuntimeDescriptor

    /// Platform identifier reported by the runtime.
    public let platform: String

    /// Protocol version selected for this session.
    public let negotiatedProtocolVersion: RuntimeProtocolVersion

    /// Runtime operations available to the Host.
    public let capabilities: [RuntimeCapability]

    /// Optional namespaced session facts.
    public let extensions: RuntimeExtensionMap?

    public init(
        runtime: RuntimeDescriptor,
        platform: String,
        negotiatedProtocolVersion: RuntimeProtocolVersion,
        capabilities: [RuntimeCapability],
        extensions: RuntimeExtensionMap? = nil
    ) {
        self.runtime = runtime
        self.platform = platform
        self.negotiatedProtocolVersion = negotiatedProtocolVersion
        self.capabilities = capabilities
        self.extensions = extensions
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        runtime = try container.decode(RuntimeDescriptor.self, forKey: .runtime)
        platform = try container.decode(String.self, forKey: .platform)
        negotiatedProtocolVersion = try container.decode(
            RuntimeProtocolVersion.self,
            forKey: .negotiatedProtocolVersion
        )
        guard negotiatedProtocolVersion == .v2 else {
            throw RuntimeProtocolVersionError.unsupportedVersion(negotiatedProtocolVersion)
        }
        capabilities = try container.decode([RuntimeCapability].self, forKey: .capabilities)
        extensions = try container.decodeIfPresent(RuntimeExtensionMap.self, forKey: .extensions)
    }

    public static let runtimeMethod = RuntimeMethod.handshake

    private enum CodingKeys: String, CodingKey {
        case runtime
        case platform
        case negotiatedProtocolVersion
        case capabilities
        case extensions
    }
}

File: Implementations/Kotlin/src/main/kotlin/dev/astrolabe/protocol/negotiation/RuntimeHandshake.kt
//
//  RuntimeHandshake.kt
//  astrolabe-protocol
//
//  Created by 轩辕十四 on 2026/7/20.
//

package dev.astrolabe.protocol

import kotlinx.serialization.Serializable

/** Host product information sent when opening a protocol session. */
@Serializable
public data class RuntimeClientDescriptor(
    /** Client product name. */
    public val name: String,
    /** Client release version. */
    public val version: String
)

/** Parameters for the handshake request. */
@Serializable
public data class RuntimeHandshakeParameters(
    /** Host initiating the protocol session. */
    public val client: RuntimeClientDescriptor,
    /** Protocol versions the Host can decode. */
    public val supportedProtocolRange: RuntimeProtocolRange
) {
    public companion object {
        /** Typed request contract for the handshake method. */
        public val contract: RuntimeMethodContract<RuntimeHandshakeParameters> by lazy {
            RuntimeMethodContract(RuntimeMethod.handshake, serializer())
        }
    }
}

/** Runtime implementation information returned by handshake. */
@Serializable
public data class RuntimeDescriptor(
    /** Namespaced runtime implementation identifier. */
    public val identifier: RuntimeNamespacedIdentifier,
    /** Runtime SDK release version. */
    public val version: String,
    /** Opaque identifier for this runtime process instance. */
    public val instanceID: RuntimeOpaqueIdentifier
)

/** Successful handshake response payload. */
@Serializable
public data class RuntimeHandshakePayload(
    /** Runtime implementation participating in the session. */
    public val runtime: RuntimeDescriptor,
    /** Platform identifier reported by the runtime. */
    public val platform: String,
    /** Protocol version selected for this session. */
    public val negotiatedProtocolVersion: RuntimeProtocolVersion,
    /** Runtime operations available to the Host. */
    public val capabilities: List<RuntimeCapability>,
    /** Optional namespaced session facts. */
    public val extensions: RuntimeExtensionMap? = null
) {
    init {
        require(negotiatedProtocolVersion == RuntimeProtocolVersion.V2) {
            "Handshake selected an unsupported protocol version"
        }
    }

    public companion object {
        /** Typed success-payload contract for the handshake method. */
        public val contract: RuntimeMethodContract<RuntimeHandshakePayload> by lazy {
            RuntimeMethodContract(RuntimeMethod.handshake, serializer())
        }
    }
}
